feat(docs-infra): set up Google Analytics 4 along with keeping legacy Universal Analytics

devversion opened the pull request:
We currently use Universal Analytics. This is deprecated in favor of
Google Analytics 4 and UA will stop processing hits in October 2023.

This change intends to prepare us for this migration, and to already
pre-populate our GA4 property (there is no way to migrate existing data
/properties into a GA4 property -- a new one needs to be created).

This will help us minimize the data gap so that we can:

* Continue to look at the UA property with the full time span until
  October 2023
* Can start using the GA4 property long-term in the future, starting
  with data even before Universal Analytics stops processing new data.

We need to keep the existing `analytics.js` setup. Initially we have
considered using `gtag.js` for both the UA and GA4 properties, as it
supports that, but that doesn't work with our strict trusted types
enforcement because it results in multiple `gtag.js` scripts (specific
versions for UA or GA4) that recreate the same trusted type policies.
This causes runtime errors and breaks the setup.

Instead, with continued use of `analytics.js` we have the benefit of
a good separation of trusted types + events and configuration. There is
some problematic with translation of Universal Analytics Events to GA4,
or the other way around (even though we don't use custom events
currently)

We also do not need to send page views for our GA4 property because GA4
with gtag supports this automatically (respecting the history state --
using the `Enhanced measurement events` setting in the UI).
For our UA legacy instance we continue to dispatch events manually. This
logic can be removed in the future.

More details can be found here:
https://docs.google.com/document/d/1aK8u4ZlXbqQ2wMqmgSX7Ces8iLgamC13oCoG6VeBruA/edit?usp=sharing&resourcekey=0-EVe-Rhnme3bj_pkz2RcOmw.

devversion committed:
feat(docs-infra): set up Google Analytics 4 along with keeping legacy Universal Analytics

We currently use Universal Analytics. This is deprecated in favor of
Google Analytics 4 and UA will stop processing hits in October 2023.

This change intends to prepare us for this migration, and to already
pre-populate our GA4 property (there is no way to migrate existing data
/properties into a GA4 property -- a new one needs to be created).

This will help us minimize the data gap so that we can:

* Continue to look at the UA property with the full time span until
  October 2023
* Can start using the GA4 property long-term in the future, starting
  with data even before Universal Analytics stops processing new data.

We need to keep the existing `analytics.js` setup. Initially we have
considered using `gtag.js` for both the UA and GA4 properties, as it
supports that, but that doesn't work with our strict trusted types
enforcement because it results in multiple `gtag.js` scripts (specific
versions for UA or GA4) that recreate the same trusted type policies.
This causes runtime errors and breaks the setup.

Instead, with continued use of `analytics.js` we have the benefit of
a good separation of trusted types + events and configuration. There is
some problematic with translation of Universal Analytics Events to GA4,
or the other way around (even though we don't use custom events
currenlty)

We also do not need to send page views for our GA4 property because GA4
with gtag supports this automatically (respecting the history state --
using the `Enhanced measurement events` setting in the UI).
For our UA legacy instance we continue to dispatch events manually. This
logic can be removed in the future.

More details can be found here:
https://docs.google.com/document/d/1aK8u4ZlXbqQ2wMqmgSX7Ces8iLgamC13oCoG6VeBruA/edit?usp=sharing&resourcekey=0-EVe-Rhnme3bj_pkz2RcOmw.

mary-poppins commented:
You can preview 73e995d at https://pr46716-73e995d.ngbuilds.io/.

gkalpak commented on aio/angular.json:
```diff
@@ -137,7 +137,8 @@
             "headers": {
-              "Content-Security-Policy": "require-trusted-types-for 'script'; trusted-types angular angular#bundler angular#unsafe-bypass aio#analytics google#safe"
+              // Keep in sync with the `firebase.json` file.
+              "Content-Security-Policy": "require-trusted-types-for 'script'; trusted-types angular angular#bundler angular#unsafe-bypass aio#analytics google#safe goog#html; report-uri https://csp.withgoogle.com/csp/angular.io"
```
Is `report-uri` needed here? Do we need to report violations in dev mode?

gkalpak commented on aio/src/app/shared/analytics.service.ts:
```diff
@@ -0,0 +1,93 @@
+
+import { environment } from '../../environments/environment';
+import { WindowToken } from 'app/shared/window';
+import { unwrapScriptUrlForSink, scriptUrl } from 'safevalues';
```
Nit: Order imported identifiers alphabetically 😇

```suggestion
import { scriptUrl, unwrapScriptUrlForSink} from 'safevalues';
```

gkalpak commented on aio/src/app/shared/analytics.service.ts:
```diff
@@ -0,0 +1,93 @@
+interface WindowWithGtag extends Window {
+  dataLayer?: any[];
+  gtag?(...args: any[]): void;
+  ga?(...args: any[]): void;
```
Does `ga` come from `gtag` or is it from the legacy analytics?

gkalpak commented on aio/src/app/shared/analytics.service.ts:
```diff
@@ -0,0 +1,93 @@
+ * Note: Presupposes that the legacy `analytics.js` script has been loaded on the
+ * host web page.
+ *
+ * Associates data with properties determined the environment configurations:
```
```suggestion
 * Associates data with properties determined from the environment configurations:
```

gkalpak commented on aio/src/app/shared/analytics.service.ts:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  private _gtag(...args: any[]) {
+    const gtagFn = (this.window as any).gtag;
```
Is the `as any` cast necessary now that `this.window` is typed as `WindowWithGtag`?

gkalpak commented on aio/src/app/shared/analytics.service.ts:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  private _legacyGa(...args: any[]) {
+    const gaFn = (this.window as any).ga;
```
Is the `as any` cast necessary now that `this.window` is typed as `WindowWithGtag`?

gkalpak commented on aio/src/app/shared/analytics.service.ts:
```diff
@@ -0,0 +1,93 @@
+      scriptUrl`https://www.googletagmanager.com/gtag/js?id=${environment.googleAnalyticsId}`;
+
+    // skip `gtag` for Protractor e2e tests.
+    if (window.name.includes('NG_DEFER_BOOTSTRAP')) {
```
Will the window name still include `NG_DEFER_BOOTSTRAP` by the time this service is instantiated?
(I don't recall how Protractor works :grin:)

devversion replied:
I checked the source code and made threw error here (tests stopped working). Seems to work.

gkalpak commented on aio/src/app/shared/analytics.service.ts:
```diff
@@ -0,0 +1,93 @@
+
+    // Note: This cannot be an arrow function as `gtag.js` expects an actual `Arguments`
+    // instance with e.g. `callee` to be set. Do not attempt to change this and keep this
+    // as much as possible in sync with the original tracking code snippet.
```
What does the "original tracking code snippet" refer to? 😕

devversion replied:
I clarified that with more comment lines

gkalpak reviewed (approved):
A couple of questions. Otherwise lgtm :+1:

_EDIT: I realized there are more than a couple of questions, but still all pretty minor_ 😛

mary-poppins commented:
You can preview fcd3d70 at https://pr46716-fcd3d70.ngbuilds.io/.

devversion committed:
test: update payload size goldens for AIO after analytics change

Most of the increase comes from code being moved from `index.html`
to the main bundle. A small increase comes from the additional `gtag.js`
loading logic, but other than that the size is effectively increased
due to the untracked code from `index.html` being moved into the
`main.js` bundle.

mary-poppins commented:
You can preview 4d39d28 at https://pr46716-4d39d28.ngbuilds.io/.

devversion committed:
refactor: move error reporting for analytics to service

This is acceptable because `gtag.js` is also only loaded as part of the
service, similary `ga` is only registered in the service (so the
reporting was a noop if errors are thrown before -- no change here)

We can benefit from downleveling etc. and do not need to be careful
with the inline script in the `index.html`

mary-poppins commented:
You can preview bee4458 at https://pr46716-bee4458.ngbuilds.io/.

mary-poppins commented:
You can preview 567b8c8 at https://pr46716-567b8c8.ngbuilds.io/.

mary-poppins commented:
You can preview b0718eb at https://pr46716-b0718eb.ngbuilds.io/.

mary-poppins commented:
You can preview b3c0a68 at https://pr46716-b3c0a68.ngbuilds.io/.

mary-poppins commented:
You can preview 342a3ed at https://pr46716-342a3ed.ngbuilds.io/.

gkalpak commented on aio/src/app/shared/analytics.service.ts:
```diff
@@ -0,0 +1,113 @@
+
+    // do not load the library when we are running in an e2e protractor test.
+    if (this.isProtractor) {
+      return;
```
Could we move this higher up?

gkalpak reviewed (approved):
One minor comment.
Otherwise lgtm :+1:

dylhunn reviewed (approved):
reviewed-for: size-tracking

devversion committed:
refactor(docs-infra): enable ip anonymization for legacy universal analytics

In GA4 ip anonymization is enabled by default, but in universal
analytics with `analytics.js`, this needs to be explicitly enabled.

mary-poppins commented:
You can preview 6d0d48f at https://pr46716-6d0d48f.ngbuilds.io/.

mary-poppins commented:
You can preview 6a0131c at https://pr46716-6a0131c.ngbuilds.io/.

jessicajaniuk commented:
@devversion This PR has conflicts with `14.0.x`. I've updated to `target: minor`. Please create a patch PR if you'd like to land in `14.0.x`.
